From 7e37afb59f42cdc2d585184294d7f50f397b966e Mon Sep 17 00:00:00 2001
From: Gleb Mazovetskiy <glex.spb@gmail.com>
Date: Wed, 13 Jan 2021 19:10:54 +0000
Subject: [PATCH 1/4] cmake: Link OSG plugins for OSG_STATIC build

Previously these plugins were not linked.
`--whole-archive` is used because plugins use registration.
---
 CMakeLists.txt             |  2 --
 apps/openmw/CMakeLists.txt | 17 +++++++----------
 2 files changed, 7 insertions(+), 12 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 4a615e843..22be162af 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -52,7 +52,6 @@ set(CMAKE_MODULE_PATH ${CMAKE_SOURCE_DIR}/cmake/)
 
 if (ANDROID)
     set(CMAKE_FIND_ROOT_PATH ${OPENMW_DEPENDENCIES_DIR} "${CMAKE_FIND_ROOT_PATH}")
-    set (OSG_PLUGINS_DIR CACHE STRING "")
 endif()
 
 # Version
@@ -289,7 +288,6 @@ if(OSG_STATIC)
     add_definitions(-DOSG_LIBRARY_STATIC)
 
     find_package(OSGPlugins REQUIRED COMPONENTS ${USED_OSG_PLUGINS})
-    list(APPEND OPENSCENEGRAPH_LIBRARIES ${OSGPlugins_LIBRARIES})
 endif()
 
 set(BOOST_COMPONENTS system filesystem program_options iostreams)
diff --git a/apps/openmw/CMakeLists.txt b/apps/openmw/CMakeLists.txt
index f96ddb27f..d35d7c4a3 100644
--- a/apps/openmw/CMakeLists.txt
+++ b/apps/openmw/CMakeLists.txt
@@ -143,25 +143,22 @@ target_link_libraries(openmw
     components
 )
 
-if (ANDROID)
-    set (OSG_PLUGINS
-        -Wl,--whole-archive
-    )
+if(OSG_STATIC)
+    set(_osg_plugins -Wl,--whole-archive)
     foreach(PLUGIN_NAME ${USED_OSG_PLUGINS})
-        set(OSG_PLUGINS ${OSG_PLUGINS} ${OSG_PLUGINS_DIR}/lib${PLUGIN_NAME}.a)
+        set(_osg_plugins ${_osg_plugins} ${OSGPlugins_LIB_DIR}/lib${PLUGIN_NAME}.a)
     endforeach()
+    set(_osg_plugins ${_osg_plugins} -Wl,--no-whole-archive)
+    target_link_libraries(openmw ${_osg_plugins})
+endif(OSG_STATIC)
 
-    set (OSG_PLUGINS
-        ${OSG_PLUGINS} -Wl,--no-whole-archive
-    )
-
+if (ANDROID)
     target_link_libraries(openmw
         EGL
         android
         log
         dl
         z
-        ${OPENSCENEGRAPH_LIBRARIES}
         freetype
         jpeg
     	png
-- 
2.27.0

